### PR TITLE
[SERVICE-436] Change production pipeline to require user interaction to deploy to NPM.

### PR DESCRIPTION
**Pull request by MichaelMCoates**
Split the production pipeline into two stages - one that deploys to CDN and one that deploys to NPM. In the NPM stage, we take in a user input to choose whether or not we deploy.

### CODE REVIEW

**Comment by MichaelMCoates** [comment]
To test on staging: 



stage('Build & Deploy (Staging)') {
            agent { label 'linux-slave' }
            when { branch "develop" }
            environment {
                GIT_SHORT_SHA = sh ( script: "git rev-parse --short HEAD", returnStdout: true ).trim()
                PKG_VERSION = sh ( script: "node -pe \"require('./package.json').version\"", returnStdout: true ).trim()

                BUILD_VERSION = "PKG_VERSION" + "-alpha." + "env.BUILD_NUMBER"
                CHANNEL = "staging"
                SERVICE_NAME = "fdc3"
                MANIFEST_NAME = "app.staging.json"

                S3_LOC = "env.DSERVICE_S3_ROOT" + "SERVICE_NAME" + "/" + "BUILD_VERSION"
                DOCS_CHANNEL_LOC = "env.DSERVICE_S3_ROOT_DOCS" + "SERVICE_NAME" + "/" + "CHANNEL"
                DOCS_VERSIONED_LOC = "env.DSERVICE_S3_ROOT_DOCS" + "SERVICE_NAME" + "/" + "BUILD_VERSION"
                MANIFEST_LOC = "env.DSERVICE_S3_ROOT" + "SERVICE_NAME" + "/" + "MANIFEST_NAME"
            }
            stages {
                stage("Build & Deploy to CDN (Staging)") {
                    steps {
                        sh "npm i --ignore-scripts"
                        sh "SERVICE_VERSION=${BUILD_VERSION} npm run build"
                        sh "echo ${GIT_SHORT_SHA} > ./dist/SHA.txt"
                        sh "npm run zip"
                        sh "npm run docs"
                        sh "aws s3 cp ./res/provider ${S3_LOC}/ --recursive"
                        sh "aws s3 cp ./dist/provider ${S3_LOC}/ --recursive"
                        sh "aws s3 cp ./dist/client/openfin-" + SERVICE_NAME + ".js ${S3_LOC}/"

                        sh "aws s3 cp ./dist/docs ${DOCS_CHANNEL_LOC} --recursive"
                        sh "aws s3 cp ./dist/docs ${DOCS_VERSIONED_LOC} --recursive"
                        sh "aws s3 cp ./dist/provider/app.json ${MANIFEST_LOC}"
                    }
                }
                stage("Deploy Client to NPM (Staging)") {
                    input {
                        message "Would you like to deploy the client to NPM?"
                        parameters {
                            choice(name: 'DEPLOY_CLIENT', choices: ['Yes', 'No'], description: '') 
                        }
                    }
                    when {expression { DEPLOY_CLIENT == 'Yes' }}
                    steps {
                        withCredentials([string(credentialsId: "NPM_TOKEN_WRITE", variable: 'NPM_TOKEN')]) {
                            sh "echo //registry.npmjs.org/:_authToken=$NPM_TOKEN > $WORKSPACE/.npmrc"
                        }
                        echo "publishing pre-release version to npm: " + BUILD_VERSION
                        sh "npm version --no-git-tag-version " + BUILD_VERSION
                        sh "npm publish --tag alpha"
                        sh "npm version --no-git-tag-version " + PKG_VERSION
                    }
                }
            }
        }

File: Jenkinsfile [hunk 1]
```diff
@@ -78,39 +78,53 @@ pipeline {
...
-                sh "aws s3 cp ./dist/provider ${S3_LOC}/ --recursive"
-                sh "aws s3 cp ./dist/client/openfin-" + SERVICE_NAME + ".js ${S3_LOC}/"
-
-                sh "aws s3 cp ./dist/docs ${DOCS_CHANNEL_LOC} --recursive"
-                sh "aws s3 cp ./dist/docs ${DOCS_VERSIONED_LOC} --recursive"
-                sh "aws s3 cp ./dist/provider/app.json ${MANIFEST_LOC}"
+                S3_LOC = "env.DSERVICE_S3_ROOT" + "SERVICE_NAME" + "/" + "BUILD_VERSION"
+                DOCS_CHANNEL_LOC = "env.DSERVICE_S3_ROOT_DOCS" + "SERVICE_NAME" + "/" + "CHANNEL"
+                DOCS_VERSIONED_LOC = "env.DSERVICE_S3_ROOT_DOCS" + "SERVICE_NAME" + "/" + "BUILD_VERSION"
+                MANIFEST_LOC = "env.DSERVICE_S3_ROOT" + "SERVICE_NAME" + "/" + "MANIFEST_NAME"
```
Comment (pjbroadbent) [inline Jenkinsfile]: Do these concatenations work as expected? Looks like `MANIFEST_LOC` would come out as the string `"env.DSERVICE_S3_ROOTSERVICE_NAME/MANIFEST_NAME"`?..

File: Jenkinsfile [hunk 2]
```diff
@@ -19,7 +19,7 @@ pipeline {
                     }
                 }
 
-                stage('Integation Tests') {
+                stage('Integration Tests') {
```
Comment (bryangaleopenfin) [inline Jenkinsfile:22]: 😃👍

**Review by bryangaleopenfin — APPROVED** [review]
Looks good, just running again on Jenkins because I'm curious how the user actually inputs their choice

File: Jenkinsfile [hunk 3]
```diff
@@ -78,39 +78,53 @@ pipeline {
...
 
-                withCredentials([string(credentialsId: "NPM_TOKEN_WRITE", variable: 'NPM_TOKEN')]) {
-                    sh "echo //registry.npmjs.org/:_authToken=$NPM_TOKEN > $WORKSPACE/.npmrc"
+                        sh "aws s3 cp ./dist/docs ${DOCS_CHANNEL_LOC} --recursive"
+                        sh "aws s3 cp ./dist/docs ${DOCS_VERSIONED_LOC} --recursive"
+                        sh "aws s3 cp ./dist/provider/app.json ${MANIFEST_LOC}"
+                    }
+                }
+                stage("Deploy Client to NPM (Production)") {
+                    input {
```
Comment (pjbroadbent) [inline Jenkinsfile:113]: Can you use `when` blocks on these?

I'm assuming the prompt would happen at this stage in the pipeline, after the tests have ran. If it's a trivial change, would be good to have the prompt right at the start of the build. Could move the `input` to the top of the script, with it's own `when` condition?..

File: Jenkinsfile (hunk 1 above)
Comment (pjbroadbent) [reply Jenkinsfile]: Looking at the test input, seems this is happening: https://builds.openfin.co/blue/organizations/jenkins/HadoukenIO%2Ffdc3-service/detail/PR-55/6/pipeline

Can you un-quote the variables, or try this syntax:
`MANIFEST_LOC = "${env.DSERVICE_S3_ROOT}${SERVICE_NAME}/${MANIFEST_NAME}"`

Comment (MichaelMCoates) [reply Jenkinsfile]: Ahhhh strange, I thought I had that working properly beforehand. Changing over to the string interpolation

File: Jenkinsfile (hunk 3 above)
Comment (MichaelMCoates) [reply Jenkinsfile:113]: Ahhh yep, it'd be nice to set this as soon as you push versus having to wait around until the tests finish. Moving.

**Review by pjbroadbent — APPROVED** [review]

**Comment by pjbroadbent** [comment]
There's some overlap with SERVICE-398 here. Going to merge now and handle the conflicts as part of that story.

**Pull request merged**